source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,1118 @@
+
+    Or nested:
+
+        [alice]
+        Q = {t = 1, m = [{q = {t = 2, m = [{n = "bob"}, {n = "carol"}]}},
+                         {q = {t = 2, m = [{n = "dave"}, {n = "elsie"}]}}]}
+
+    Params:
+        quorums = the quorum set
+        level = indentation level
+
+*******************************************************************************/
+
+private string toToml (QuorumConfig[PublicKey] quorums)
+{
+    return quorums.byKeyValue.map!(pair =>
+        format("[%s]\n%s", pair.key.pretty, toToml(pair.value)))
+            .join("\n\n");
+}
+
+/// ditto
+private string toToml (QuorumConfig config, size_t level = 0)
+{
+    string result;
+
+    result ~= format("%s = {t = %s, m = [",
+        level == 0 ? "Q" : "q",
+        config.threshold);
+
+    // nodes
+    auto nodes = config.nodes
+        .map!(qnode => format(`{n = "%s"}`, qnode.pretty))
+        .join(", ");
+
+    result ~= nodes.to!string;
+
+    // subquorums (recursive)
+    if (config.quorums.length > 0)
+    {
+        result ~= ", ";
+        auto subq = config.quorums
+            .map!(qsub => format("{ %s}", toToml(qsub, level + 1)))
+            .join(", ");
+
+        result ~= subq.to!string;
+    }
+
+    result ~= "]}";
+    return result;
+}
+
+version (unittest)
+{
+    /// prettyr formatting for [sets of] public keys
+    private auto pretty (Set!PublicKey input)
+    {
+        return input._set.byKey.map!(key => key.pretty);
+    }
+}
+else
+{
+    private auto pretty (PublicKey input)
+    {
+        return input;
+    }
+}
+
+/// The rest is support code for testing, quorum prettify routines, etc.
+version (unittest):
+
+/// converts "GA..." => n0 / n1, etc for easier debugging
+private string[PublicKey] key_to_simple;
+
+static this ()
+{
+    foreach (node_idx; 0 .. pregen_seeds.length)
+        key_to_simple[getTestKey(node_idx)] = format("n%s", node_idx);
+}
+
+/// ditto
+private string pretty (QuorumConfig[PublicKey] conf, size_t seed_idx)
+{
+    auto keys = conf.keys;
+    sort!((a, b) => key_to_simple[a] < key_to_simple[b])(keys);
+
+    string res;
+    auto hash = hashFull(seed_idx);
+    res ~= format("\nrand idx: %s (seed: %s)\n", seed_idx, hash.prettify);
+
+    foreach (key; keys)
+    {
+        auto quorum = conf[key];
+        auto nodes = quorum.nodes.map!(node => key_to_simple[node]).array;
+        sort(nodes);
+
+        res ~= format("\n%s: QuorumConfig(thresh: %s, nodes: %s)",
+            key_to_simple[key],
+            quorum.threshold,
+            nodes);
+    }
+
+    return res;
+}
+
+/// ditto
+private string pretty (QuorumConfig conf)
+{
+    auto nodes = conf.nodes.map!(node => key_to_simple[node]).array;
+    return format("\nQuorumConfig(%s, %s)",
+        conf.threshold,
+        nodes);
+}
+
+// convenience so we can use [n0, n1] when refering to public keys in the tests
+static foreach (node_idx; 0 .. pregen_seeds.length)
+    mixin("PublicKey n%s;".format(node_idx));
+
+#line __LINE__
```

Comment:
Please tell me you have a bug report for this ?

Reply:
I did this out of memory because I seem to remember mixins messing with line numbers. But that was years ago. I'll try to remove it.

Reply:
It does, but I don't know the details of how / when